### DISCUSSION TITLE
Why is Forward inversely in Matrix? Is it design or bug?

### DISCUSSION THREAD

**Discussion by a995049470** [stride3d/stride#1754] [Q&A] [answered]
# Question and/or Comment

Matrix.cs
![image](https://user-images.githubusercontent.com/37652153/227925256-78058292-d3b9-469d-99fc-1f0c8286b792.png)

**Comment by NicusorN5** [comment, marked as answer]
By design, Stride uses a right handed coordinate system, that uses -Z as forward.

**Comment by NicusorN5** [comment]
Can anyone close this thread, or move it to a discussion? Thanks.

**Reply by VaclavElias** [reply]
I migrated it.

**Reply by NicusorN5** [reply]
Thanks.

**Comment by Thaina** [comment]
Actually should be a bug. But it was disguised as *legacy* and/or *industry standard*  with no real purpose except continuing to propagate the norm of architect since the time they have drawn blueprint on the table

**Reply by TranquilAbyss** [reply]
My suggestion is a setting to flip this so that game designer to decide how this is implemented. Stride can keep the default to right hand rule then but let new developer decide which one to use.

I got caught off guard with this as well. Coming from Unity I thought I found a bug but decided to check here first. I would prefer it changed. but I think that would be difficult since others have already built on top of this setup. My understand is that 3D modeling software and engineering is done right and rule but cameras and game engines are the other ways. It would be nice to see this changed, but for now I will have to flip the Z axis client side for my implementation .

**Comment by TranquilAbyss** [comment]
I got a message via email asking how is "different coordinate system a bug?". A bug is not just broken code that prevents software to work. This coordinate system change, for some of us, could be classify this as a Logical Bug. Those of us who are familiar with a different coordinate system it affects our workflow, like badly named variable can confuse developer and create problems, Logical Bugs include assigning a value to the wrong variable. Which to those not familiar with the right handed coordinate system the Code and "wrong variable" being subjective see this as the negative numbers assigned to the wrong variable.

Changing this would not be a huge issues since require some dependency injection of two different classes that handle left and right handed matrices. You can see it as an unnecessary quality of life change but we see it as a disruptive Logical bug that can create friction with our workflow.

**Reply by IXLLEGACYIXL** [reply]
every engine uses a different one and is up to the engine designers
![coordinate-comparison-chart-full-799565922](https://github.com/stride3d/stride/assets/107197024/e03b425b-203d-413d-a46b-2fc1d691810b)

**Reply by Thaina** [reply]
@IXLLEGACYIXL I would argue only LH+YUp and RH+ZUp is right ones. Another two is abomination just for the sake of compatibility

Rationally, we already always have positive direction in 3 dimension. Up and forward is ingrain in our perspective. And left to right is the traditionally majority direction of written human language. So all 3 direction should be positive according to this reasoning

What's really arguable is which one is Y and which one is Z. And it's understandable that architecture which is the source of the need for 3D file format would choose Z to be up from the table. Because they work on the table which is horizontal plane and upward is only sensible Z direction

But now we all work on screen which is vertical plane. X and Y direction is rationally be used for the plane itself. And so Z need to be depth dimension

There are no reason we need to keep the table coordinate on the screen we use, except for the convenient of file format compatibility. So I am saying it was a bug because it similar to Y2K bug that we just pick the wrong format from the start. In case you don't remember, before year 2000 many data and database just have 2 digit number so it start making ambiguous data after 100 years

3D file format is the same. It's a wrong choice that we just choose from habit and convenience. Now it's bite back that it was not intuitive and make unnecessary fragmentation

**Reply by TranquilAbyss** [reply]
Thanks for the info. I disagree that in modern times its up to the engine designers, since Unity and Unreal has larger communities, not because of coordinate system but its ingraining an opinion. Its up to the game designers to stay or go from an engine. Making Stride more seamless is a good marketing move to keep people looking for alternatives to stay. I am still weighting the pros and cons of Stride, I get mix messaging from the communities acceptance of new ideas.

In my opinion for my project, its a logic bug. Understand that something can be a bug and not a bug based on the context. I built my server code before I choosing Stride and I have not plan on changing the server code. I understand staying with Stride may not change any part of this. This image does not make this issue any less of a bug for my project for me to workaround. If I swap out Game Engines coordinate system may consider later.

If you want to see the Stride community grow and I would recommend adding features that are compatible with the two popular game engines.
